File: DFS.py
class DFS:
    def __init__(self, n):
        self.N = n
        self.G = [[False for x in range(self.N)] for y in range(self.N)]
        self.Acc = [False] * self.N
        self.sta = []
        self.setupGraph()
        print("---------- The DFS, as follows --------------------")

    def DFS_(self, start, g):
        at = 0
        addToSta = " "
        self.sta.append(start)
        self.Acc[start] = True
        print(f"\t\tThe stack: [{self.retCity(self.sta[-1])}...top=>]\n")

        while True:
            at = self.sta.pop(-1)
            if g == at:
                print(f"\nThe goal ( {self.retCity(g)} )has been reached\n\n")
                return

            print(f"-->At {self.retCity(at)} city now ")
            addToSta = f"{self.retCity(at)} has not been the goal,children ["

            for i in range(self.N):
                if not self.Acc[i] and self.G[at][i]:
                    self.sta.append(i)
                    addToSta += f"{self.retCity(i)} "
                    self.Acc[i] = True
            addToSta += "]'v been added to the stack"
            print(addToSta)

            print("The stack: [ ", end="")
            for i in self.sta:
                print(f"{self.retCity(i)} ..", end="")
            print("top=>]\n")
            if not self.sta:
                break

    def setupGraph(self):
        self.G[1][0] = self.G[0][1] = True
        self.G[0][5] = self.G[5][0] = True
        self.G[0][6] = self.G[6][0] = True
        self.G[1][2] = self.G[2][1] = True
        self.G[1][3] = self.G[3][1] = True
        self.G[1][6] = self.G[6][1] = True
        self.G[1][9] = self.G[9][1] = True
        self.G[2][3] = self.G[3][2] = True
        self.G[2][4] = self.G[4][2] = True
        self.G[2][6] = self.G[6][2] = True
        self.G[3][9] = self.G[9][3] = True
        self.G[4][6] = self.G[6][4] = True
        self.G[4][7] = self.G[7][4] = True
        self.G[5][6] = self.G[6][5] = True
        self.G[5][7] = self.G[7][5] = True
        self.G[6][7] = self.G[7][6] = True
        self.G[7][8] = self.G[8][7] = True

    def retCity(self, i):
        if i == 0:
            return "Business and Economics"
        elif i == 1:
            return "Science"
        elif i == 2:
            return "Engineering"
        elif i == 3:
            return "Architecture and Planning"
        elif i == 4:
            return "Sharia and Islamic Studies"
        elif i == 5:
            return "Medicine"
        elif i == 6:
            return "Qassim University Public Library"
        elif i == 7:
            return "Department of English & Translation"
        elif i == 8:
            return "Parking lot 1"
        else:
            return "Parking lot 2"


def main():
    ob = DFS(10)
    print("\n\nChoose a city to start with(its number): \n")
    for i in range(10):
        print(f"{ob.retCity(i)} city[{i}]")
    cityChoice = int(input("\nInput: "))
    print("\nChoose a city to be the goal(its number): ")
    goal = int(input("\nInput: "))

    if cityChoice < 0 or cityChoice >= 10:
        print("Mistake,run the program again")
        return

    ob.DFS_(cityChoice, goal)


if __name__ == "__main__":
    main()

File: java/DFS.java
package java;

import java.util.Scanner;
import java.util.*;

public class DFS {
    int N; // number of vertices in the graph
    boolean[][] G; // the graph as an adjacency matrix
    // G[i][j] is true if there is an edge from i to j
    boolean[] Acc; // a nodes takes a true value when it is visited or added to the queue
    Stack<Integer> sta; // G[i][j] is true if there is an edge from i to j

    public DFS(int size) {
        N = size;
        Acc = new boolean[N]; // a visited array to mark which vertices have been visited while doing the DFS
        sta = new Stack<Integer>();
        setupGraph(size);
        System.out.println("---------- The DFS, as follows --------------------");
        System.out.println();

    }

    // perform a DFS starting at node start
    public void DFS(int start, int g) {
        int at;
        String addToSta = " ";
        sta.push(start);
        Acc[start] = true;

        System.out.print("\t\tThe stack: ");
        System.out.println("[" + retCity(sta.peek()) + "...top=>]\n");

        do {
            at = sta.pop();

            if (g == at) {
                System.out.println("\nThe goal (" + retCity(g) + ")has been reached\n\n");
                return;
            }

            System.out.println("-->At " + retCity(at) + " city now ");
            addToSta = retCity(at) + " has not been the goal,children [";

            for (int i = 0; i < N; ++i)
                if (!Acc[i] && G[at][i]) {
                    sta.push(i);
                    addToSta += retCity(i) + " ";
                    Acc[i] = true;
                }
            addToSta += "]'v been added to the stack";
            System.out.println(addToSta);

            System.out.print("The stack: [");
            for (int i : sta)
                System.out.print(retCity(i) + "..");
            System.out.println("top=>]\n");

        } while (!sta.empty());
    }

    public void setupGraph(int size) {
        // set up a graph with a number of vertices:

        G = new boolean[N][N];

        G[0][1] = G[1][0] = True
        G[0][5] = G[5][0] = True
        G[0][6] = G[6][0] = True
        G[1][2] = G[2][1] = True
        G[1][3] = G[3][1] = True
        G[1][6] = G[6][1] = True
        G[1][9] = G[9][1] = True
        G[2][3] = G[3][2] = True
        G[2][4] = G[4][2] = True
        G[2][6] = G[6][2] = True
        G[3][9] = G[9][3] = True
        G[4][6] = G[6][4] = True
        G[4][7] = G[7][4] = True
        G[5][6] = G[6][5] = True
        G[5][7] = G[7][5] = True
        G[6][7] = G[7][6] = True
        G[7][8] = G[8][7] = True
    }

    public static void main(String[] args) {
        int cityChoice, goal;
        Scanner inp = new Scanner(System.in);
        System.out.println("\n\nChoose a city to start with(its number): \n");
        for (int i = 0; i < 10; i++)
            System.out.println(retCity(i) + " city[" + i + "]");
        System.out.print("\nInput: ");
        cityChoice = inp.nextInt();
        System.out.println("\nChoose a city to be the goal(its number): ");
        System.out.print("\nInput: ");
        goal = inp.nextInt();

        if (cityChoice < 0 || cityChoice >= 10) {
            System.out.println("Mistake,run the program again");
            System.exit(0);
        }

        DFS schDFS = new DFS(10);
        schDFS.DFS(cityChoice, goal);

    }

    public static String retCity(int i) // the function returns city name, according to its index in V array
    {
        if (i == 0)
            return "Business and Economics";
        else if (i == 1)
            return "Science";
        else if (i == 2)
            return "Engineering";
        else if (i == 3)
            return "Architecture and Planning";
        else if (i == 4)
            return "Sharia and Islamic Studies";
        else if (i == 5)
            return "Medicine";
        else if (i == 6)
            return "Qassim University Public Library";
        else if (i == 7)
            return "Department of English & Translation";
        else if (i == 8)
            return "Parking lot 1";
        else
            return "Parking lot 2";
    }
}